res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ *
+ *
+ * This library depends on Lodash, which is copyright JS Foundation
+ * and other contributors and licensed under the MIT license. Refer to
+ * the lodash.mixxx.js file in this directory for details.
+**/
+
+(function(global) {
+    const Component = class {
+        constructor(options) {
+            Object.assign(this, options);
+            // These cannot be in the prototype; they must be unique to each instance.
+            this.connections = [];
+            this.isShifted = false;
+
+            if (options !== undefined && typeof options.key === "string") {
+                this.inKey = options.key;
+                this.outKey = options.key;
+            }
+
+            if (typeof this.unshift === "function") {
+                this.unshift();
+            }
+
+            if (this.outConnect !== false && this.group !== undefined && this.outKey !== undefined) {
+                this.connect();
+                if (this.outTrigger !== false) {
+                    this.trigger();
+                }
+            }
+        }
+        // default methods
+        // You should probably overwrite at least some of these.
+        inValueScale(value) {
+            // Hack to get exact center of pots to return 0.5
+            if (value > (this.max / 2)) {
+                return (value - 1) / (this.max - 1);
+            } else {
+                return value / (this.max + 1);
+            }
+        }
+        // map input in the XML file, not inValueScale
+        input(channel, control, value, _status, _group) {
+            this.inSetParameter(this.inValueScale(value));
+        }
+        outValueScale(value) { return value * this.max; }
+        output(value, _group, _control) {
+            this.send(this.outValueScale(value));
+        }
+
+        // common functions
+        // In most cases, you should not overwrite these.
+        inGetParameter() {
+            return engine.getParameter(this.group, this.inKey);
+        }
+        inSetParameter(value) {
+            engine.setParameter(this.group, this.inKey, value);
+        }
+        inGetValue() {
+            return engine.getValue(this.group, this.inKey);
+        }
+        inSetValue(value) {
+            engine.setValue(this.group, this.inKey, value);
+        }
+        inToggle() {
+            this.inSetValue(!this.inGetValue());
+        }
+
+        outGetParameter() {
+            return engine.getParameter(this.group, this.outKey);
+        }
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
```

Comment:
the block of functions above says they should not be overridden, but this one says it should.  Can we order the function definitions so override-ok functions are all on top, and the others below?  That way things most useful to the controller programmer are all in one place.